Add SweepFormula

ukos-git opened the pull request:
Add formula serialization. The formula style follows [json logic](http://jsonlogic.com/)

closes #238

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+					continue
 				endif
-				JSON_AddJSON(jsonID, jsonPath, FormulaParser(formula[-inf, i - 1]))
-				JSON_AddJSON(jsonID, jsonPath, FormulaParser(formula[i + 1, inf]))
```
this was limiting to binary operations

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
 			case "*":
-				if(strsearch(formula, "+", 0) > -1)
+				if((strsearch(formula, "+", i) != -1) || (strsearch(formula, "-", i) != -1))
+					str += token // look ahead for conflicts
```
I need the look ahead to respect formulas like

```
print JSON_DUMP(FormulaParser("5*1+2*3+4+5*10"), indent = 2)
```

```json
{
  "+": [
    {
      "*": [
        5,
        1
      ]
    },
    {
      "*": [
        2,
        3
      ]
    },
    4,
    {
      "*": [
        5,
        10
      ]
    }
  ]
}
```

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+
+		// state
 		strswitch(token)
+			case "(":
```
these states are currently not in effect

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+		// path action
+		if(!!cmpstr(state, lastState))
+			if(!cmpstr(state, "+") && !cmpstr(lastState, "*") && !cmpstr(jsonPath, "/" + lastState))
+				print "not working"
```
actually, now it is working, so this is a left-over

ukos-git replied:
I added the positive look ahead in L46

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+			endif
+
+			if(JSON_GetType(jsonID, jsonPath) == JSON_ARRAY)
+				JSON_AddObjects(jsonID, jsonPath)
```
This is the usual `JSON_AddTreeArray` issue where the automatic creation of objects is not allowed.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+		else
+			JSON_AddString(jsonID, jsonPath, str)
+		endif
+	elseif(strlen(str) > 0)
```
handle last element (no associated operation, but belongs to the last object of the current state)

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -12,41 +12,117 @@ Function FormulaParser(formula)
+		if(!cmpstr(num2str(str2num(formula)), formula))
+			JSON_AddVariable(jsonID, jsonPath, str2num(formula))
+		else
+			JSON_AddString(jsonID, jsonPath, str)
```
in case it is a string like for function arguments, add the string.

ukos-git commented:
@t-b mostly clean commits. We had the review until 2f13091 last time. Please take a look at the unit tests. If you do not see what a line does, remove it and check, which unit test fails. Documenting the state machine other than with the working unit tests is nearly impossible.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -437,3 +523,75 @@ Function/WAVE GetSweepForFormula(graph, rangeStart, rangeEnd, channelType, sweep
+			mainPanel = GetMainWindow(ba.win)
+			DFREF dfr = BSP_GetFolder(mainPanel, MIES_BSP_PANEL_FOLDER)
+			SVAR formula = dfr:sweepFormula
+			FormulaPlotter(mainPanel, formula, dfr = dfr)
```
and calls the formula plotter

timjarsky commented on Packages/MIES/MIES_BrowserSettingsPanel.ipf:
```diff
@@ -996,33 +1003,56 @@ Window BrowserSettingsPanel() : Panel
+	ValDisplay status_sweepFormula_parser,limits={-1,1,0},barmisc={0,0},mode= 1,highColor= (0,65535,0),lowColor= (0,0,0),zeroColor= (65535,0,0)
+	ValDisplay status_sweepFormula_parser,value= #"NaN"
+	ValDisplay status_sweepFormula_parser,userdata(tabnum)=  "5",userdata(tabcontrol)=  "Settings"
+	Button button_sweepFormula_display,pos={52.20,88.20},size={37.80,18.00},proc=button_sweepFormula_display,title="Display"
```
@ukos-git I'm not sure why I see this:

![image](https://user-images.githubusercontent.com/5693697/64442930-a50f9380-d085-11e9-9be8-f6f9c6a68a36.png)

ukos-git replied:
Did I link the procedure file "MIES_SweepFormula.ipf"  in the "MIES_include" file? That could be it.

timjarsky replied:
@ukos-git you are correct; MIES_SweepFormula is missing. The JSON XOP is also required, correct?

ukos-git replied:
right. The JSON XOP is parsing the formula into JSON logic. It will presumably also help with our messy WaveNote storage functionality. I am not sure how to integrate it in the install procedure of MIES, though, and the introduced changes will also explicitly require IP8.

ukos-git replied:
you will need the XOP and the json_functions.ipf

timjarsky replied:
> right. The JSON XOP is parsing the formula into JSON logic. It will presumably also help with our messy WaveNote storage functionality. I am not sure how to integrate it in the install procedure of MIES, though, and the introduced changes will also explicitly require IP8.

@t-b could you help here? I guess we'll need to set up a new submodule for the JSON XOP?

t-b replied:
@timjarsky Yes, we need to intergrate the JSON-XOP as well. I'll do that together with the IP8-only move.

t-b replied:
@timjarsky Installer needs an adaptation as well.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -437,3 +523,75 @@ Function/WAVE GetSweepForFormula(graph, rangeStart, rangeEnd, channelType, sweep
+	return 0
+End
+
+Function button_sweepFormula_display(ba) : ButtonControl
```
maybe button name and procedure name could be named differently to see if that produces the conflict but another solution could be that the procedure file "MIES_SweepFormula.ipf" is not included in the "MIES_include" file

t-b commented:
@ukos-git Regarding our earlier conversation regarding how to test the `sweep` operation. An alternative approach would be to use a replacement function for `PA_GetTraceInfos` (just define an override function in the builtin procedure window). In that way you can create just a graph and test the sweep operation without the requirement to really do data acquisition. The `RunAllTests.pxp` does something like that already.

t-b commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -0,0 +1,597 @@
+	String str
+
+	return ReplaceString("/", str, "~1")
+End
```
Can it be that this function should actually be part of the JSON XOP?

ukos-git replied:
The function only escapes names of the elements. So I think, it is the user's decision where the path ends and where the element starts. Besides that, that is very primitive escaping and straight forward imho. So probably no.

t-b replied:
Okay fine. Remember you can pass `esc=0/1` to `JSON_GetKeys` to tune the escaping.

ukos-git replied:
[JSON_GetKeys](https://docs.byte-physics.de/json-xop/operations.html#get-keys) seems to be escaping by default. Should I turn that off using esc=1? - It is probably more readable turned off.

ukos-git replied:
sorry. It does not escape by default.

ukos-git replied:
I think it will not get prettier using the `esc` option:

```igorpro
	ASSERT(JSONtype == JSON_OBJECT, "Topmost element needs to be an object")
	WAVE/T operations = JSON_GetKeys(jsonID, jsonPath, esc = 1)
	ASSERT(DimSize(operations, ROWS) == 1, "Only one operation is allowed")
	jsonPathOld = jsonPath
	jsonPath += "/" + operations[0]
	WAVE/T operations = JSON_GetKeys(jsonID, jsonPathOld, esc = 0)
```

ukos-git replied:
and btw, syntax highlighting of such snippets is not working.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -253,42 +253,71 @@ Function/WAVE FormulaExecutor(jsonID, [jsonPath])
+		numIndices = DimSize(indices, ROWS)
+		if(numIndices == 1 && DimSize(out, ROWS) == 1)
+			return FormulaExecutor(jsonID, jsonPath = jsonPath + "/0", graph = graph) // function
+		elseif(numIndices > 0)
```
@t-bo data alignment not defined. Add new data to last dimension. Thus, the top array (`out`) will be the last (!) dimension if an operation needs to be carried out. The array needs to get dynamically expaneded to +1 dimension.

timjarsky commented:
[simple_analysis_example.zip](https://github.com/AllenInstitute/MIES/files/3628743/simple_analysis_example.zip)
@t-b @ukos-git I've attached a zipped PXP that includes some hacky code (in the experiment procedure file) for doing analysis on DB sweeps. It outputs 3 plots. I'd like to be able to replicate the analysis in the PXP using this PR. Please ignore any fitting errors (they are not relevant). Sorry!

t-b commented:
@ukos-git I've started diving into the code some more. I've added some commits on top. The two WIPs should be self explanatory. And I found that "-1"/"+1" does not work, I've added failing tests for these.

More from me later.

ukos-git commented:
It will be hard to implement `-1` but give it a try. I cannot think of a clever solution to this right now. Probably need to check if the buffer is empty if an operation is encountered and send that to `STATE_HIGHER_ORDER`.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -411,37 +426,100 @@ Function/WAVE FormulaExecutor(jsonID, [jsonPath, graph])
+
+			strswitch(dimension[0])
+				case "x":
+					SetScale x, offset[0], delta[0], unit[0], data
```
weired bug. Probably wrong data usages.

Use setscale(range(0…10), 0, 1)

@t-b @MichaelHuth 

If `delta[0] == 1` or any other unsigned integer, I can use `print delta[0]` to output 1.
So print outputs the right thing but the wave is scaled to negative values like `-1.6` for 1 and `-1.7` for 5 if I remember correctly. That could be some wrongly interpreted data format. I stored the wave data in a variable and it behaves similar. Also using str2num(num2str()) to clean up gives same result.

MichaelHuth replied:
> If you do not use the /P flag, SetScale converts num1  and num2  into a starting index value and a delta index value

Are you sure this should scale on the index values?

ukos-git replied:
I have added the `/P` flag and now it works for `ROWS` but `COLS` is still broken.

I did not quite get what the difference between omitting `/P` and explicitly giving `/P` is.

t-b replied:
Tracked in #295.

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -677,8 +677,10 @@ Function/WAVE FormulaExecutor(jsonID, [jsonPath, graph])
+			ASSERT(DimSize(wv, LAYERS) <= 1, "Unhandled dimension")
+			ASSERT(DimSize(wv, CHUNKS) <= 1, "Unhandled dimension")
+			MatrixOP/FREE out = log(wv)
+			FormulaWaveScaleTransfer(wv, out, COLS, ROWS)
```
can it be `lg` instead to indicate base 10

t-b replied:
Rename to `log10`.

t-b commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -810,7 +810,7 @@ Function/WAVE GetSweepForFormula(graph, rangeStart, rangeEnd, channelType, chann
 
 	// we want the sweeps sorted with ascending sweep numbers
-	SortColumns/A/DIML/KNDX={FindDimLabel(traces, COLS, "sweepNumber"} sortWaves=traces
+	SortColumns/A/DIML/KNDX={FindDimLabel(traces, COLS, "sweepNumber")} sortWaves=traces
```
Thx. I forgot to push on saturday.

t-b commented:
Required issues which need to be resolved for this PR can be seen [here](https://github.com/AllenInstitute/MIES/issues?q=is%3Aissue+is%3Aopen+label%3Amvp).

ukos-git commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -70,6 +70,8 @@ Function FormulaParser(formula)
 	Variable jsonID = JSON_New()
 	String jsonPath = ""
 
+	formula = LowerStr(formula)
```
note that this will convert also `AD0` to `ad0`. If your intention was to allow case-insensitive operations/function naming then this should better go directly to the executor, wrapping `Lowerstr(operation[0])` and we should not camelCase operations then.

ukos-git replied:
dd89f1802c4717af66f3eec109c4b5348d5c331c

t-b commented:
@ukos-git Could you rebase against latest master the next time you rebase? Thanks.

t-b commented on Packages/MIES/MIES_SweepFormula.ipf:
```diff
@@ -783,50 +783,31 @@ Function FormulaPlotter(graph, formula, [dfr])
 		ModifyGraph/W=$win mode=0
 	endif
+
+	RestoreCursors(win, cursorInfos)
```
Thx.

t-b commented on Packages/doc/.gitignore:
```diff
@@ -16,4 +16,3 @@ ZeroMQ-XOP-Readme.rst
 backup_*.dot
 *.zip
 *.dot.svg
-sphinx-output.log
```
Yes that is the idea. But I did not want to rebase and step on your toes.

t-b commented:
@ukos-git I always get an assertion when I open the databrowser in a fresh experiment. Do you see that as well?

t-b commented:
BSP_GetFolder does not return a valid dfr.

t-b commented:
@ukos-git How do I edit the Help notebook?

t-b commented:
@ukos-git I've pushed all I have for today.

ukos-git commented:
Also pushed my changes. Rebasing would give two conflicts that are autoresolvable by `kdiff3`.

t-b commented:
@ukos-git Got for the rebase!

ukos-git commented:
your test for differentiate needs an update.

```
  Assuming equality using mode WAVE_DATA for waves _free_ and _free_: is false  Assertion "REQUIRE_EQUAL_WAVES(output, input, mode = WAVE_DATA)" failed in line 561, procedure "UTF_SweepFormula.ipf"
  Leaving test case "UTF_SweepFormula#testDifferentiales
```

ukos-git commented:
Regarding the notebook:

I have edited the text in a separate notebook and copy/pasted it.

t-b commented:
Test failure is tracked in https://github.com/AllenInstitute/MIES/issues/294#issue-497719994.

ukos-git commented:
good to go.

timjarsky commented:
@t-b 
```
max(data(
	cursors(C,D),
	channels(AD),
	range(0,9,1)
))
-
avg(data(
	cursors(A,B),
	channels(AD),
	range(0,9,1)
))
```

(1) gives a yFormula with 18 rows (0 through 17). I would have expected 20 rows.

(2) The sweep formula plot is empty. I expected yFormula vs calculated.

(3) The JSON tab and the Help tab are not hidden when selecting a tab in the DB other then SF.

![image](https://user-images.githubusercontent.com/5693697/65568231-def9ea00-df0d-11e9-9d68-e874a6723aec.png)

![image](https://user-images.githubusercontent.com/5693697/65568264-fa64f500-df0d-11e9-972c-fd92fb5b0b2a.png)

timjarsky commented:
@t-b 
```
max(data(
	cursors(C,D),
	channels(AD),
	range(0,9,1)
))
-
avg(data(
	cursors(A,B),
	channels(AD),
	range(0,9,1)
))

vs

max(data(
	cursors(E,F),
	channels(DA4),
	range(0,9,1)
))
``` 
gives:
` ** WAVE gave error: WAVE reference to "out" failed.`

I also tried channels(DA, 4) but that gave:

```
  Please provide the following information if you contact the MIES developers:
  ################################
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Stacktrace:
button_sweepFormula_display(...)#L947 [MIES_SweepFormula.ipf]
FormulaPlotter(...)#L749 [MIES_SweepFormula.ipf]
FormulaExecutor(...)#L409 [MIES_SweepFormula.ipf]
FormulaExecutor(...)#L359 [MIES_SweepFormula.ipf]
FormulaExecutor(...)#L690 [MIES_SweepFormula.ipf]
GetSweepForFormula(...)#L824 [MIES_SweepFormula.ipf]
PA_GetTraceInfos(...)#L248 [MIES_PulseAveraging.ipf]
GetAllSweepTraces(...)#L3352 [MIES_MiesUtilities.ipf]
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Time: 2019-09-24T21:10:52-07:00
  Locked device: [- none -]
  Current sweep: [- none -]
  DAQ: [- none -]
  Testpulse: [- none -]
  Experiment: Untitled.pxp
  Igor Pro version: 8.0.4.1 (33900)
  MIES version:
  Release_2.0_20190602-242-g5747147d
Date and time of last commit: 2019-09-24T15:23:05+02:00
Submodule status: 
160000 550cea2cd1e7e77cb7b7396bb9a536f805f6a889 0	Packages/IPNWB
160000 7b6d2115733523463bb9d0250cee38a41a88d8b3 0	Packages/ITCXOP2
160000 aa1858d915ee28ebf5974f59f0bc19a421e5c28e 0	Packages/ZeroMQ
160000 da79879e19e6467daa58fb075f8428ae15df2061 0	Packages/doc/doxygen-filter-ipf
160000 a097151fbc9a8a8f2e504c9e46adcb327a19c8f3 0	Packages/unit-testing
  ################################
```

ukos-git commented:
> (1) gives a yFormula with 18 rows (0 through 17). I would have expected 10 rows and two columns.

please see the [definition for the range function](https://www.w3schools.com/python/ref_func_range.asp). It is a bit inconvenient at first, but it makes sense that `range(0,9,1) == range(9)` outputs 9 elements and not 10.

> (2) The sweep formula plot is empty. I expected yFormula vs calculated (for each column).

I am getting this:

![FormulaPlot](https://user-images.githubusercontent.com/16167358/65619259-1de87800-dfc0-11e9-8b41-28ce41242726.png)

> (3) The JSON tab and the Help tab are not hidden when selecting a tab in the DB other then SF.

Igor Pro had issues with the tabcontrol, but I also forgot to link the `ACL` tabcontrol procedure.

> (4) The help tab is great! It may not be clear to users how an array relates to a sweep. Can we help them out?

Every function argument is an array element. Every wave is an array. This means that basically everything is an array.

> gives:
> ** WAVE gave error: WAVE reference to "out" failed.

One of your statements returned invalid. I updated the procedure to return `NaN` and not `$""` here. Maybe that is more clear. debug could be improved with proper error messages. As we want to have a reversed data object in #272 the error would go to the corresponding function and we could highlight the errorness part in the window. But that should be separate from here.

>I also tried channels(DA, 4) but that gave:

With this statement you are selecting two datasets: all `DA` channelTypes and all channelNumbers matching `4` in the third dimension. The data sizes probably missmatch if you select the channelNumbers like this, so I would not do that at all.

I also added plotting against text waves if that helps and cleaned up some commits from yesterday. 

I will be off for 1 week. I have a few more urgent tasks to tackle. Maybe @t-b can take over.

timjarsky commented:
@t-b could the user-provided `lo` and `hi` filter frequencies for the Butterworth filter be in Hz?

t-b commented:
@timjarsky Should not crash anymore with the butterworth filter. I had to force the latest nightly for it.

t-b commented:
@timjarsky @ukos-git Rebased onto latest master (which is finally IP8 only).

timjarsky commented:
@t-b formula that was working now opens debugger. Let me know if you want to screen share or need a copy of the pxp.

![image](https://user-images.githubusercontent.com/5693697/66707497-42e62600-ecf6-11e9-9a89-bab27811d0b1.png)

![image](https://user-images.githubusercontent.com/5693697/66707491-2d70fc00-ecf6-11e9-9820-0c5f10f7c1a2.png)

t-b commented:
@timjarsky My bad, I see the error from the code in the debugger already. Will be fixed.

t-b commented:
@timjarsky Fixed.

t-b commented:
Note to self: Needs rebase and integration of new controls with resize capability.

timjarsky commented:
@t-b should the Butterworth filter be observable in the displayed traces?

![image](https://user-images.githubusercontent.com/5693697/66718993-fa367780-ed9e-11e9-93dc-b7a7b2b06d4e.png)

t-b commented:
@timjarsky Yes. Have you already switched to Hz for the lower and upper bound?

timjarsky commented:
https://github.com/AllenInstitute/MIES/pull/245#issuecomment-541439444

No; I'm using 1/hz. Is that right?

t-b commented:
@timjarsky You can just pass Hz into `butterworthfilter`. I've now rebased against master and now the subpanel is resizable as well ;)

t-b commented:
@timjarsky Thanks for the review and approving. One GUI control procedure was not valid and therefore the tests failed. I've fixed that now.
